lib/runtime.js

```diff
@@ -127,7 +64,70 @@ exports.enable = function (Module) {
     return exports[name];
   };
 
-  return Module;
+  // If key is provided, it will be used to identify the given setters so
+  // that they can be replaced if module.importSync is called again with the
+  // same key. This avoids potential memory leaks from import declarations
+  // inside loops. The compiler generates these keys automatically (and
+  // deterministically) when compiling nested import declarations.
+  module.importSync = function (id, setters, key) {
+    setESModule(this);
+
+    var absoluteId = this.resolve(id);
+
+    if (! (typeof setters === "object" && setters !== null)) {
+      this.require(absoluteId);
+      return;
+    }
+
+    var entry = Entry.getOrCreate(absoluteId);
+    entry.addSetters(this, setters, key);
+
+    var countBefore = entry.runCount;
+    var exports = this.require(absoluteId);
+
+    if (entry.runCount === countBefore) {
+      // If require(absoluteId) didn't run any setters for this entry,
+      // perhaps because it's not the first time this module has been
+      // required, run the setters now using an object that passes as the
+      // real module object.
+      entry.runModuleSetters({
+        id: absoluteId,
+        exports: exports,
+        getExportByName: this.getExportByName
+      });
+    }
+  };
+
+  // Platform-specific code should find a way to call this method whenever
+  // the module system is about to return module.exports from require. This
+  // might happen more than once per module, in case of dependency cycles,
+  // so we want Module.prototype.runModuleSetters to run each time.
+  module.runModuleSetters = function (valueToPassThrough) {
+    var entry = Entry.get(this.id);
+    if (entry !== null) {
+      entry.runModuleSetters(this);
+    }
+
+    // Assignments to exported local variables get wrapped with calls to
+    // module.runModuleSetters, so module.runModuleSetters returns the
+    // valueToPassThrough parameter to allow the value of the original
+    // expression to pass through. For example,
+    //
+    //   export var a = 1;
+    //   console.log(a += 3);
+    //
+    // becomes
+    //
+    //   module.export("a", () => a);
+    //   var a = 1;
+    //   console.log(module.runModuleSetters(a += 3));
+    //
+    // This ensures module.runModuleSetters runs immediately after the
+    // assignment, and does not interfere with the larger computation.
+    return valueToPassThrough;
+  };
```

Comment:
Since these instance methods only refer to `this` and not `module`, I think we can define them at the top level (outside of the `enable` function) and just copy them onto the `module` instance.